### PR TITLE
feat: new apis to conditionally append or set headers

### PR DESCRIPTION
**Pull request by mumairofficial**
## PR Checklist
Please check if your PR fulfills the following requirements:

- [x] The commit message follows our guidelines: https://github.com/angular/angular/blob/master/CONTRIBUTING.md#commit
- [x] Tests for the changes have been added (for bug fixes / features)
- [x] Docs have been added / updated (for bug fixes / features)

## PR Type
What kind of change does this PR introduce?

- [ ] Bugfix
- [x] Feature
- [ ] Code style update (formatting, local variables)
- [ ] Refactoring (no functional changes, no api changes)
- [ ] Build related changes
- [ ] CI related changes
- [ ] Documentation content changes
- [ ] angular.io application / infrastructure changes
- [ ] Other... Please describe:

## What is the current behavior?

User has to write conventional logic such as conditional statements to add extra headers and so on.

Issue Number: N/A

## What is the new behavior?
These two new apis in `HttpHeader` `appendIf` and `setIf` will help to reduce conditional
logic code on end user's development environment specially in `HttpInterceptors` (i.e. only add token if it some condition is true). 

Only set or append new header value if condition flag is True otherwise return current `HttpHeader` instance state.

Pros: Less lines of code, code is predictable, readable and easy to maintained 

### Example: 

**Before**
```ts
@Injectable()
export class MyHttpInterceptor implements HttpInterceptor {

  constructor(private store: MyStore) {}

  // override
  intercept(req: HttpRequest<any>, next: HttpHandler): Observable<HttpEvent<any>> {
    const isLoggedIn = this.auth.isLoggedIn(); // true/false

    const requestWithDefaultHeaders = req.clone({
        headers: req.headers
          .set('Content-Type', 'application/json')
          .set('X-RM-DATA', 'Value')
    });

    // condition to append token
    let requestWithRequiredHeaders = requestWithDefaultHeaders.clone()

    if (isLoggedIn) {
        requestWithRequiredHeaders = requestWithDefaultHeaders.clone({
        // confusion ? requestWithDefaultHeaders or req
        headers: requestWithDefaultHeaders.headers  
          .append('Authorization', `Bearer ${this.auth.token}`)
      });
    }

    // if I only check for account then it will add empty header to request
    // i.e. { Account: <ampty> } into request
    if (isLoggedIn && this.store.account) { 
      requestWithRequiredHeaders = requestWithRequiredHeaders.clone({
        headers: requestWithRequiredHeaders.headers
          .append('Account', this.store.account)
      })
    }

    // only append when redirected from xyz launchpad
    if (!isLoggedIn && this.store.xyzLaunchPadSecret) {
      requestWithRequiredHeaders = requestWithDefaultHeaders.clone({
        // confusion ? 
        // requestWithDefaultHeaders or req or requestWithRequiredHeaders
        headers: requestWithDefaultHeaders.headers  
          .append('XYZ-SECRET', this.store.xyzLaunchPadSecret)
          .append('XYZ-CLIENT-CHALLENGE', this.store.xyzLaunchPadClientChallenge)
      });
    }

    // ... few more headers based upon some conditional logic (omitted on purpose)

    if (isLoggedIn) {
      return next.handle(requestWithRequiredHeaders)
    } else if (!isLoggedIn && this.store.xyzLaunchPadSecret) {
      return next.handle(requestWithRequiredHeaders)
    } else {
      return next.handle(requestWithDefaultHeaders)
    }
  }
}
```

**After**
```ts
@Injectable()
export class MyHttpInterceptor implements HttpInterceptor {

  constructor(private store: MyStore) {}

  // override
  intercept(req: HttpRequest<any>, next: HttpHandler): Observable<HttpEvent<any>> {
    const requestWithHeaders = req.clone({
        headers: req.headers
          .append('Content-Type', 'application/json')
          .append('X-RM-DATA', 'Value')
          // if following conditions are not true. it will not polute the request with blank headers
          .appendIf(this.isLoggedIn(), 'Authorization', `Bearer ${this.auth.token}`)
          .appendIf(this.isAuthorizedAndAccount(), 'Account', this.store.account)
          .appendIf(this.isRedirectedFromLaunchPad(), 'XYZ-SECRET', this.store.xyzLaunchPadSecret)
          .appendIf(this.isRedirectedFromLaunchPad(), 'XYZ-CLIENT-CHALLENGE', this.store.xyzLaunchPadClientChallenge)
    });

    return next.handle(requestWithHeaders);
  }

  private isLoggedIn(): boolean {
    // return this.auth.isLoggedIn();
  }
  
  private isAuthorizedAndAccount(): boolean {
    // this.isLoggedIn() && some logic
  }

  private isRedirectedFromLaunchPad(): boolean {
    // return this.isLoggedIn && some logic
  }
}
```

## Does this PR introduce a breaking change?

- [ ] Yes
- [x] No

### CODE REVIEW

**Comment by mumairofficial** [comment]
@googlebot I signed it!

**Comment by mumairofficial** [comment]
@googlebot I fixed it.

File: packages/common/http/src/headers.ts [hunk 1]
```diff
@@ -159,6 +177,25 @@ export class HttpHeaders {
   set(name: string, value: string|string[]): HttpHeaders {
     return this.clone({name, value, op: 's'});
   }
+
+  /**
+   * Conditionally sets a header value for a given name. If flag is true and If the header name
```
Comment (ayazhafiz) [inline packages/common/http/src/headers.ts]:
```suggestion
   * Conditionally sets a header value for a given name. If flag is true and if the header name
```

File: packages/common/http/src/headers.ts [hunk 2]
```diff
@@ -159,6 +177,25 @@ export class HttpHeaders {
...
+   * already exists,
+   * its value is replaced with the given value.
```
Comment (ayazhafiz) [inline packages/common/http/src/headers.ts]: can be unified on one line

**Comment by dannyskoog** [comment]
IMHO I think adding `setIf` and `appendIf` doesn't bring that much value to be frank. They're just assignment functions with a guard.

Also - adding these would be a precedent for introducing more of these `if`-suffixed functions. E.g. `FormControl.setValueIf()`, etc.

File: packages/common/http/src/headers.ts (hunk 2 above)
Comment (mumairofficial) [reply packages/common/http/src/headers.ts]: Suggested formatting is done

**Comment by mumairofficial** [comment]
@dannyskoog Although they are simple methods with guard clauses but In my opinion some of these methods are good to have; specially when use-case is like object configuration (i.e. decorator) because they provides more readability and control in declarative programming fashion.

**Comment by ayazhafiz** [comment]
I agree with Danny. I don't this adds too much value; in your example, consider the case that generating a token is somewhat expensive, so you'd want to generate a token only if a user is logged in. Then you would have to provide a guard anyway, and there is no use for `appendIf`. I think such scenarios are likely to be more common than the scenarios where `appendIf` would be most useful.

In general factory methods like `append` and `set` used to create common objects with well-defined properties. `appendIf` is definitely not unhelpful in certain scenarios, but I think that it is better to leave the control flow involved in creating a request to the creator of that request rather than asking the library to do so.

**Comment by mumairofficial** [comment]
@ayazhafiz @dannyskoog Gotcha ... seems like my example was very simple. I kinda agree with you guys but here is situation from where I went through and feel like these two can be life saver.

> Angular is framework and it's not a language and Frameworks = Save Life

> In other languages overloading is possible but in TypeScript we wish (that's why -if post fix the methods)

Let me explain from enterprise application in which I am working right now; I happen to add an header in interceptor but when I open up source file I kinda confused although debugging bail me out.

My example may sound subjective but here is small code snippet (only some figures are changed rest is 100% same): there are three more headers I just omit to generate the feel
```ts
@Injectable()
export class MyHttpInterceptor implements HttpInterceptor {

  constructor(private store: MyStore) {}

  // override
  intercept(req: HttpRequest<any>, next: HttpHandler): Observable<HttpEvent<any>> {
    const isLoggedIn = this.auth.isLoggedIn(); // true/false

    const requestWithDefaultHeaders = req.clone({
        headers: req.headers
          .set('Content-Type', 'application/json')
          .set('X-RM-DATA', 'Value')
    });

    // condition to append token
    let requestWithRequiredHeaders = requestWithDefaultHeaders.clone()

    if (isLoggedIn) {
        requestWithRequiredHeaders = requestWithDefaultHeaders.clone({
        // confusion ? requestWithDefaultHeaders or req
        headers: requestWithDefaultHeaders.headers  
          .append('Authorization', `Bearer ${this.auth.token}`)
      });
    }

    // if I only check for account then it will add empty header to request
    // i.e. { Account: <ampty> } into request
    if (isLoggedIn && this.store.account) { 
      requestWithRequiredHeaders = requestWithRequiredHeaders.clone({
        headers: requestWithRequiredHeaders.headers
          .append('Account', this.store.account)
      })
    }

    // only append when redirected from xyz launchpad
    if (!isLoggedIn && this.store.xyzLaunchPadSecret) {
      requestWithRequiredHeaders = requestWithDefaultHeaders.clone({
        // confusion ? 
        // requestWithDefaultHeaders or req or requestWithRequiredHeaders
        headers: requestWithDefaultHeaders.headers  
          .append('XYZ-SECRET', this.store.xyzLaunchPadSecret)
          .append('XYZ-CLIENT-CHALLENGE', this.store.xyzLaunchPadClientChallenge)
      });
    }

    // ... few more headers based upon some conditional logic (omitted on purpose)

    if (isLoggedIn) {
      return next.handle(requestWithRequiredHeaders)
    } else if (!isLoggedIn && this.store.xyzLaunchPadSecret) {
      return next.handle(requestWithRequiredHeaders)
    } else {
      return next.handle(requestWithDefaultHeaders)
    }
  }
}
```
Did you noticed the confusion part what to clone and which request headers to clone. also difficult to dig in!

After use of appendIf()

```ts
@Injectable()
export class MyHttpInterceptor implements HttpInterceptor {

  constructor(private store: MyStore) {}

  // override
  intercept(req: HttpRequest<any>, next: HttpHandler): Observable<HttpEvent<any>> {
    return req.clone({
        headers: req.headers
          .append('Content-Type', 'application/json')
          .append('X-RM-DATA', 'Value')
          // if following conditions are not true. it will not polute the request with blank headers
          .appendIf(this.isLoggedIn(), 'Authorization', `Bearer ${this.auth.token}`)
          .appendIf(this.isAuthorizedAndAccount(), 'Account', this.store.account)
          .appendIf(this.isRedirectedFromLaunchPad(), 'XYZ-SECRET', this.store.xyzLaunchPadSecret)
          .appendIf(this.isRedirectedFromLaunchPad(), 'XYZ-CLIENT-CHALLENGE', this.store.xyzLaunchPadClientChallenge)
    });

  }

  private isLoggedIn(): boolean {
    // return this.auth.isLoggedIn();
  }
  
  private isAuthorizedAndAccount(): boolean {
    // this.isLoggedIn() && some logic
  }

  private isRedirectedFromLaunchPad(): boolean {
    // return this.isLoggedIn && some logic
  }
}
```

There is big difference without magic trick! Consider how many developers time is saved; 
In after example my code is predictable, readable even my colleague knows how he can add extra header with some criteria.

Framework = Save Life :)

**Comment by ayazhafiz** [comment]
You're definitely right that there are cases where declarative conditional methods like `appendIf` could be convenient - for example, exactly the example you provided. That's true, and I also think that it is not the job of the `HttpHeaders` API to solve architectural control flow complexities. `HttpHeaders` is little more than simply a store of HTTP headers, and it is intentional to separate application control flow from such a storage.

The worry for me is that `appendIf` complicates the API surface of `HttpHeaders` and solves use cases that are not all that common. I think that most clients, when they need to create different header sets for different conditions, will create separate those headers in separate blocks anyway. This makes the distinction of unique headers explicit and modular.

For your application, if `appendIf` and `setIf` are critical, you could write a wrapper:

```typescript
class ConditionalHeaders {
    constructor(public headers: HttpHeaders);
    appendIf(...);
    setIf(...);
}
```

Though I make no value statement about whether this is a good or bad practice. All I wish to say is that I don't think it's helpful for the `HttpHeaders` library to have `appendIf` or `setIf` methods at this time.

**Comment by mumairofficial** [comment]
@ayazhafiz You cleared so many things :-) and appendIf usecases are limited and btw I already did refactoring by wrapping the HttpHeaders creation but thank you for suggestion

**Comment by ayazhafiz** [comment]
Sure :-)! Feel free to close this issue if the conversation is resolved.

**Pull request closed without merging**